packages/lib/src/components/Dropin/Dropin.tsx

```diff
@@ -82,14 +82,27 @@ class DropinElement extends UIElement<DropinElementProps> {
             return this.activePaymentMethod.updateWithAction(action);
         }
 
+        const pmType = action.paymentMethodType || 'scheme';
+
         // Extract desired props that we need to pass on from the pmConfiguration for this particular PM
-        const pmConfig = getComponentConfiguration(action.paymentMethodType, this.props.paymentMethodsConfiguration);
+        const pmConfig = getComponentConfiguration(pmType, this.props.paymentMethodsConfiguration);
+
+        // action.type === 'threeDS2Fingerprint' creates backwards compatibility with old API versions
+        const threeDS2Options =
+            action.type === 'threeDS2' || action.type === 'threeDS2Fingerprint'
+                ? {
+                      elementRef: this.elementRef,
+                      ...(this.props.challengeWindowSize && { challengeWindowSize: this.props.challengeWindowSize }),
+                      ...(pmConfig.challengeWindowSize && { challengeWindowSize: pmConfig.challengeWindowSize })
+                  }
+                : null;
```

Comment:
Could this be moved into `src/core/ProcessResponse/PaymentAction/actionTypes.ts`?

Reply:
How? - these are all props & config specific to the Dropin element that need to be passed on in order to configure the action component

Reply:
On both `actionTypes.ts` and the ThreeDS2 component itself you should be able to access a prop named `isDropin` that will be injected when the component has been created on a Drop-in context. 
Using that, you could set those specific options in those places, instead of the more generic Dropin class.

Reply:
But `isDropin` is just a boolean. 
So it doesn't contain the `this` that is required to get the `elementRef` nor does it contain the `pmConfig` necessary to asses which level of `challengeWindowSize` (and further down in the code, `onError`) we need for the action. 
The problem here is less what I'm trying to do for 3DS2 and more a deeper one of which (of the 3 possible levels) of config properties we take to pass onto the action. 
And this `handleAction` function is surely where we configure and collect those properties prior to handing them on to the action.
So yes, I could find a way to do all this logic in the `actionTypes.ts` (always pass `elementRef` and `pmConfig` to the action so _it_ can do the sorting out of what it needs...) but is that really the place to do it?
Is the problem here that you feel that having the code here is _dirtying_ the `Dropin.tsx` component?

Reply:
We could just do this:
```
handleAction(action: PaymentAction) {
        if (!action || !action.type) throw new Error('Invalid Action');

        if (this.activePaymentMethod.updateWithAction) {
            return this.activePaymentMethod.updateWithAction(action);
        }

        const pmType = action.paymentMethodType || 'scheme';

        // Extract desired props that we need to pass on from the pmConfiguration for this particular PM
        const pmConfig = getComponentConfiguration(pmType, this.props.paymentMethodsConfiguration);

        const paymentAction: UIElement = this.props.createFromAction(action, {
            isDropin: true,
            onAdditionalDetails: state => this.props.onAdditionalDetails(state, this),
            onError: this.props.onError, // Add ref to onError in case the merchant has defined one in the component options
            ...(pmConfig?.onError && { onError: pmConfig.onError }), // Overwrite ref to onError in case the merchant has defined one in the pmConfig options
            elementRef: this.elementRef,
            ...(this.props.challengeWindowSize && { challengeWindowSize: this.props.challengeWindowSize }),
            ...(pmConfig.challengeWindowSize && { challengeWindowSize: pmConfig.challengeWindowSize })
        });

        if (paymentAction) {
            return this.setStatus(paymentAction.props.statusType, { component: paymentAction });
        }
        return null;
    }
```

Reply:
I think it makes sense to send down `elementRef` and `this.props.paymentMethodsConfiguration`, and then do the actual mapping in either the actionType function or the component itself.

Reply:
Now send `elementRef` and spread all `props` and `paymentMethodsConfiguration` into the options object sent to the action